File: Leetcode/GroupAnagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> ans;
        map<string, int> m;
        int now = 1;
        for(int i = 0; i < strs.size(); ++i){
            string t = strs[i];
            sort(t.begin(), t.end());
            int x = m[t];
            if(x != 0){
                ans[x - 1].push_back(strs[i]);        
            }
            else{
                m[t] = now;
                ans.push_back(vector<string>());
                ans[now - 1].push_back(strs[i]);
                ++now;
            }           
        }
        return ans;
    }
};

File: Leetcode/GroupAnagrams.rs
// https://leetcode.com/problems/group-anagrams

use std::collections::{HashMap, HashSet};

struct Solution {}

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let strs = strs
            .into_iter()
            .map(|s| {
                let mut s_ = s.clone().as_bytes().to_vec();
                s_.sort();
                (s_, s)
            })
            .collect::<Vec<_>>();
        let uq = strs.iter().map(|x| &x.0).collect::<HashSet<_>>();
        let mut uq = uq.into_iter().fold(HashMap::new(), |mut hm, x| {
            hm.entry(x).or_insert(vec![]);
            hm
        });

        for (v, s) in strs.iter() {
            uq.entry(&v).and_modify(|y| y.push(s.clone()));
        }

        uq.values().map(|x| x.clone()).collect()
    }
}
